### DISCUSSION TITLE
Toast notifications on Ubuntu?

### DISCUSSION THREAD

**Discussion by logiclrd** [AvaloniaUI/Avalonia#15914] [Q&A] [answered]
I need to show toast notifications from my app running on Ubuntu. I've seen other Ubuntu apps show toast notifications. Does Avalonia UI expose any wrapper for this functionality? As I understand it, it must be something provided natively, because Wayland doesn't permit applications to position their windows, so if the apps were simply creating their own windows to show toasts, they wouldn't appear properly docked to the edge of the screen on Wayland I think?

**Comment by thevortexcloud** [comment, marked as answer]
Avalonia does have a notification API. I don't think it shows up on the desktop though.

https://reference.avaloniaui.net/api/Avalonia.Controls.Notifications/WindowNotificationManager/

There were (are?) plans to make it show OS notifications. That said most desktops have some sort of dbus API you can call to show a notification.

https://specifications.freedesktop.org/notification-spec/notification-spec-latest.html

You are correct though, Wayland does not allow for apps to position themselves. The way programs like Steam get around this is by injecting their code directly into a game's rendering code. But if you don't need a custom notification, you should just use the dbus API. It should be more portable anyway.

https://wiki.archlinux.org/title/Desktop_notifications

**Reply by logiclrd** [reply]
Thanks very much :-) I don't really understand how the team behind Wayland can continue to refuse to allow windows to position themselves. The discussions I see all centre around the idea that Wayland could be used on devices/in environments where the very concept of positioning a window doesn't make sense. Okay, but _I_ am (hypothetically) using it on my desktop with two rectangular screens, and _I_ need apps to be able to position their windows. :-P Combined with the suggestion I keep seeing that Wayland is positioned to completely replace X11, it feels like there's a gap here... Ah well. That's all off-topic.

**Reply by thevortexcloud** [reply]
> Thanks very much :-)

You are welcome. Avalonia also already ships with tmds.dbus on Linux so you should be able to use it for this.

https://github.com/tmds/Tmds.DBus

It's a hard problem to solve for Wayland. They are trying to support everything, including non floating window managers (including some fairly esoteric ones) where it would never properly work anyway as the best thing a window manager could do in that situation is snap to the nearest valid position. Which might cause other problems if an app is expecting it to work like how it does on Windows.

> Okay, but I am (hypothetically) using it on my desktop with two rectangular screens, and I need apps to be able to position their windows.

It's not even a hypothetical problem either, there are already VR desktops which allow people to see a desktop on 360 degrees and with windows at arbitrary depths (and also because it's VR people can also physically step behind the windows too which means they would also need rotational information to display properly as well as depth). Also because it's a virtual environment, not constrained by physical dimensions it theoretically also has infinite space. 

Wayland is also not used entirely on desktops either. Things like infotainment systems in cars also run Wayland using a custom compositor and automotive Linux, where the concept of windows is generally non existent.

>  and I need apps to be able to position their windows. 

The wayland logic is you should not need to. That's not your job, it's the window managers. Since you can't possibly know how the window manager is configured or what layouts it supports. I agree though it would be nice to have something, but I am not sure what they can easily do about it.

That said the latest proposal is to instead of allowing absolute positioning is to allow for positioning in vaguely defined zones to allow for things like floating tool windows.

https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/264

**Reply by logiclrd** [reply]
What I don't understand is why they can't allow the application to provide _hints_ that it can ignore if they can't be made to make sense with the current environment, and then the _common_ case of a computer with a rectangular screen can use the hints to do what applications have always done in the past. Step one would be (x, y)x(w, h) hints, and once that system was in place, any shortcomings could be identified and hints could be made richer. To just say, "Nerp, windows shouldn't ever be prepositioned, because that doesn't work _everywhere_" doesn't feel right to me. Ah well. :-P

**Reply by logiclrd** [reply]
Just to add some additional thoughts, it definitely feels like the current approach in Wayland on this is, "If we can't provide a single solution that encompasses all paradigms at once, then we won't provide anything at all", and this does not work _at all_ for at least one of my use cases. I use a system to manage video being displayed on one or more screens/projectors. I need the ability to fire up `mpv` and say "Okay, you go to _that_ screen", and as far as I've been able to tell, I literally can't do that with Wayland. The current state of affairs has also produced things like Steam injecting code into games it is running so it can draw directly onto their windows, since Wayland doesn't provide any way for Steam to create its own overlay window positioned relative to the games. I dunno. It just feels unnecessarily puristic to me, and if Xorg goes away without Wayland providing any effective replacement for this functionality, it's going to be a problem, for me at least. :-(

**Reply by thevortexcloud** [reply]
>  "If we can't provide a single solution that encompasses all paradigms at once, then we won't provide anything at all",

Their argument is even optional features can become defacto mandatory features  as people will complain when their apps don't work in certain environments which means they will end up having to support it everywhere anyway.

> I need the ability to fire up mpv and say "Okay, you go to that screen", and as far as I've been able to tell, I literally can't do that with Wayland.

As long as the app is full screen, I am pretty sure you can actually do this. Although I don't know how myself. Full screen apps get slightly different treatment because "full screen" is a much easier semantic to handle than "go to this exact position".

If this an embedded device though the "official" way you are supposed to do this sort of thing is by either writing your own Wayland implementation/compositor or by writing a Weston extension which can then decide where to put your windows for you based on what ever rules it has. EG Weston has a Kiosk mode that is designed for a single app that runs full screen which is very close to what you want to do anyway (it also lets you specify which full screen app to place on which screen if you have multiple apps).

https://wayland.pages.freedesktop.org/weston/toc/kiosk-shell.html

There is also the ivi shell which apparently has some window management stuff, but I have no idea how it works in practice.

https://wayland.pages.freedesktop.org/weston/toc/ivi-shell.html

This talk seems to cover both shells as well as some other options in more detail:

https://youtu.be/QhGJL4nKfAg?t=286

If you are targeting generic desktops however, then there is not much you can do right now.

Part of the problem wayland devs have is people are so used to doing things the X11 way they don't think about other possible ways to solve the problem. Which is another reason why they keep rejecting stuff like absolute window positioning being in the standard.

Wayland itself does also allow for custom protocol extensions to be added. Something just has to implement it.

EDIT: I forgot about gamescope. It might also work for you and looks like it possibly lets you specify which monitor to run it on.

https://github.com/ValveSoftware/gamescope

**Reply by logiclrd** [reply]
Thanks very much :-) I still don't really agree with the approach they're taking on this particular issue, but at least I understand it a bit more, and I hadn't considered the possibility of writing my own code to get the behaviour I want. :-)

**Reply by logiclrd** [reply]
Getting back on topic, I have tried to show toasts using DBus. It evidently is successfully communicating, because it receives back a non-zero Notification Id and is able to pass it in for `replaces_id`. A test app provided by an author who made a Gist with the necessary bindings and such shows it repeatedly replacing, getting back the same ID each time. But, no toasts appear. Some final piece is missing, and I don't know what it is. Maybe it's even just configuration, where somehow my desktop environment thinks it's not supposed to show me notifications?? Not sure how to debug this.

**Reply by thevortexcloud** [reply]
I just tested it myself, and it works fine using dbus directly as well as `notify-send`. Granted I am not using Gnome, so I am not sure what's going on here. That said depending on which desktop you are using you may not even have a notification server installed. Gnome and KDE however do ship with one out of the box.

> where somehow my desktop environment thinks it's not supposed to show me notifications??

Do you have some sort of do not disturb mode enabled? I know on some desktops that will hide pretty much all notifications.

> Not sure how to debug this.

There are a few GUI and command line tools for directly interacting with dbus things. The easiest is probably `dbus-monitor` which lets you watch dbus messages.

**Reply by logiclrd** [reply]
Okay, `notify-send` works for me. Perhaps I could simply use the `notify-send` binary as a subroutine? :-P Isn't that a legitimate UNIX paradigm?

**Reply by logiclrd** [reply]
`dbus-monitor` looks super useful here, I can compare the difference between `notify-send` and whatever my app is pooping out. Thanks :-)

**Reply by logiclrd** [reply]
Ahh, found it! It was `urgency`. My application was defaulting to 0, which apparently means "not urgent at all". :-P `notify-send` was using 1. Changed mine to 1, and I'm getting notifications now! Thanks so much :-)

**Reply by thevortexcloud** [reply]
> Isn't that a legitimate UNIX paradigm?

Technically the UNIX way is "one program that does one thing and does it well". However it's fairly messy to do with .NET, which is API focused not file/program focused. Going out to an external program is always a pain in .NET. Which is why I suggested using dbus since it's much more API like and any program you call for this is likely going to use dbus anyway.

>  Thanks so much :-)

You are welcome!
